### PR TITLE
Feat/add-caller-to-earn-service

### PR DESCRIPTION
**Pull request by 0xKoaj**
Add caller parameter to buildTx EarnService methods

### CODE REVIEW

File: src/services/earn/earn-service.ts [hunk 1]
```diff
@@ -612,27 +612,25 @@ export class EarnService implements IEarnService {
...
-      }),
-    ]);
+    const strategyAddress = await this.providerService.getViemPublicClient({ chainId }).readContract({
+      address: EARN_STRATEGY_REGISTRY.address(chainId),
+      abi: strategyRegistryAbi,
+      functionName: 'getStrategy',
+      args: [strategyId],
+    });
+    const { result } = await this.providerService.getViemPublicClient({ chainId }).simulateContract({
+      address: strategyAddress,
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:622]: You need to set the `account` param, with the vault's address

File: src/services/earn/types.ts [hunk 2]
```diff
@@ -100,7 +104,7 @@ export type EarnActionSwapConfig = { slippagePercentage?: number; txValidFor?: T
 export type AddFundsEarn = { swapConfig?: EarnActionSwapConfig } & (
   | { permitData: PermitData['permitData']; signature: string }
   | { token: TokenAddress; amount: BigIntish }
-);
+) & { caller: Address };
```
Comment (0xsambugs) [inline src/services/earn/types.ts:107]: Why do we need this here? Don't we have it already as a param for deposit/increase?

**Review by 0xsambugs — COMMENTED** [review]
Great work, left some comments

File: src/services/earn/earn-service.ts [hunk 3]
```diff
@@ -738,7 +737,7 @@ export class EarnService implements IEarnService {
 
       const { result } = await this.providerService
         .getViemPublicClient({ chainId })
-        .simulateContract({ ...specialWithdrawTx, address: EARN_VAULT_COMPANION.address(chainId) });
+        .simulateContract({ ...specialWithdrawTx, address: caller as ViemAddress });
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:740]: Actually. The address should be the vault, the account should be the caller

File: src/services/earn/types.ts (hunk 2 above)
Comment (0xKoaj) [reply src/services/earn/types.ts:107]: Yes, i forgot to remove this

**Review by 0xsambugs — APPROVED** [review]

**Pull request merged**